File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/identity/provider_config.rb
# frozen_string_literal: true

module IAIso
  module Identity
    # Configuration for an OidcVerifier.
    class ProviderConfig
      attr_reader :discovery_url, :jwks_url, :issuer, :audience,
                  :allowed_algorithms, :leeway_seconds

      def initialize(discovery_url: nil, jwks_url: nil, issuer: nil, audience: nil,
                     allowed_algorithms: ["RS256"], leeway_seconds: 5)
        @discovery_url = discovery_url
        @jwks_url = jwks_url
        @issuer = issuer
        @audience = audience
        @allowed_algorithms = allowed_algorithms.map(&:to_s).freeze
        @leeway_seconds = leeway_seconds.to_i
        freeze
      end

      def self.defaults = new

      def self.okta(domain:, audience:)
        new(
          discovery_url: "https://#{domain}/.well-known/openid-configuration",
          issuer: "https://#{domain}",
          audience: audience,
        )
      end

      def self.auth0(domain:, audience:)
        new(
          discovery_url: "https://#{domain}/.well-known/openid-configuration",
          issuer: "https://#{domain}/",
          audience: audience,
        )
      end

      def self.azure_ad(tenant:, audience:, v2: true)
        base = v2 ? "https://login.microsoftonline.com/#{tenant}/v2.0"
                  : "https://login.microsoftonline.com/#{tenant}"
        new(
          discovery_url: "#{base}/.well-known/openid-configuration",
          issuer: base,
          audience: audience,
        )
      end
    end
  end
end

File: IAIso-v5.0/core/iaiso-java/iaiso-identity/src/main/java/io/iaiso/identity/ProviderConfig.java
package io.iaiso.identity;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/** Configuration for an {@link OidcVerifier}. */
public final class ProviderConfig {
    public final String discoveryUrl;       // nullable
    public final String jwksUrl;            // nullable
    public final String issuer;             // nullable — empty means trust discovery
    public final String audience;           // nullable — empty disables aud check
    public final List<String> allowedAlgorithms;
    public final long leewaySeconds;

    public ProviderConfig(String discoveryUrl, String jwksUrl, String issuer, String audience,
                          List<String> allowedAlgorithms, long leewaySeconds) {
        this.discoveryUrl = discoveryUrl;
        this.jwksUrl = jwksUrl;
        this.issuer = issuer;
        this.audience = audience;
        this.allowedAlgorithms = Collections.unmodifiableList(
            allowedAlgorithms != null ? allowedAlgorithms : Arrays.asList("RS256"));
        this.leewaySeconds = leewaySeconds;
    }

    public static ProviderConfig defaults() {
        return new ProviderConfig(null, null, null, null, Arrays.asList("RS256"), 5);
    }

    /** Build a {@link ProviderConfig} for Okta. */
    public static ProviderConfig okta(String domain, String audience) {
        return new ProviderConfig(
            "https://" + domain + "/.well-known/openid-configuration",
            null,
            "https://" + domain,
            audience,
            Arrays.asList("RS256"),
            5);
    }

    /** Build a {@link ProviderConfig} for Auth0. */
    public static ProviderConfig auth0(String domain, String audience) {
        return new ProviderConfig(
            "https://" + domain + "/.well-known/openid-configuration",
            null,
            "https://" + domain + "/",
            audience,
            Arrays.asList("RS256"),
            5);
    }

    /** Build a {@link ProviderConfig} for Azure AD / Entra. */
    public static ProviderConfig azureAd(String tenant, String audience, boolean v2) {
        String base = v2
            ? "https://login.microsoftonline.com/" + tenant + "/v2.0"
            : "https://login.microsoftonline.com/" + tenant;
        return new ProviderConfig(
            base + "/.well-known/openid-configuration",
            null,
            base,
            audience,
            Arrays.asList("RS256"),
            5);
    }
}
